Make noisy channel exclusion during Reference compatible with MATLAB PREP

a-hurst opened the pull request:
<!--Thanks for contributing-->
<!--If this is your first time, please make sure to read the contributing guideline-->
<!--https://github.com/sappelhoff/pyprep/blob/master/.github/CONTRIBUTING.md-->

# PR Description

Closes #91. This PR does three main things:

- Changes `Reference.robust_reference` to stop permanently excluding channels that are flagged as bad-by-SNR prior to initial re-referencing.
- Changes `Reference.robust_reference` to exclude bad-by-dropout channels from the full set of bad channels when`matlab_strict` is `True`, matching MatPREP's (likely accidental) behaviour.
- Adds an `as_dict` argument to `NoisyChannels.get_bads` that returns bad channels by type as a dictionary in order to facilitate the above changes. In addition to being generally useful, it also saves a lot of dict construction in `reference.py`.

I still need to update the `whats_new.rst` and `matlab_differences.rst` here, but I figured given the time-zone difference that it would be better to get the code up for review before I went to bed. Thanks in advance for checking this over!

# Merge Checklist

<!--To merge your PR we need to first take the following points into account.-->
<!--Please just leave this checklist untouched-->

- [x] the PR has been reviewed and all comments are resolved
- [x] all [CI][what-is-ci] checks pass
- [x] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [x] (if applicable): bug fixes, new features, or [API][what-is-api] changes are documented in [whats_new.rst][whats-new-file]


[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration
[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
[what-is-api]: https://en.wikipedia.org/wiki/Application_programming_interface
[whats-new-file]: https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

sappelhoff commented on pyprep/reference.py:
```diff
@@ -134,17 +134,7 @@ def perform_reference(self):
-            "bad_by_ransac": noisy_detector.bad_by_ransac,
-            "bad_all": noisy_detector.get_bads(),
-        }
+        self.noisy_channels_before_interpolation = noisy_detector.get_bads(as_dict=True)
```
this is nice :sunglasses:

sappelhoff commented on pyprep/reference.py:
```diff
@@ -260,7 +230,7 @@ def robust_reference(self):
         iterations = 0
-        noisy_channels_old = []
+        previous_bads = set()
         max_iteration_num = 4
```
perhaps adding a comment here that this number is derived from the paper? Otherwise it may appear magical

a-hurst replied:
It's actually a configurable parameter in MatPREP (4 is just the default): should we make it a parameter in PyPREP as well?

sappelhoff replied:
Yes, I think that'd be best 👍

sappelhoff commented on pyprep/reference.py:
```diff
@@ -272,51 +242,43 @@ def robust_reference(self):
+            noisy_new = noisy_detector.get_bads(as_dict=True)
+
+            # Specify bad channel types to ignore when updating noisy channels
+            # NOTE: MATLAB PREP ignores dropout channels, possibly by mistake?
```
here it would be nice if you could open that issue on matprep and link to the issue from here.

If you don't want to do that, please just link to the specific comment where this is discussed in our own issue thread :)

a-hurst replied:
Just created an issue here: https://github.com/VisLab/EEG-Clean-Tools/issues/28, will add a reference in the code.

sappelhoff reviewed (commented):
LGTM! Only minor comments for docs

a-hurst commented:
@sappelhoff Okay, I've updated the docs and whats-new, and have addressed your comments. A couple questions before this gets merged:

- Any thoughts on whether there should be an API for configuring `max_iterations` at the `PrepPipeline` level? There's currently no API for configuring `NoisyChannels` thresholds from `PrepPipeline` either, so I wasn't sure what the expectation was for `Reference` settings.
- While writing out the `matlab_differences` explanation, I realized two weird things about the current logic we might want to change:
    - With the current changes, channels flagged as bad-by-SNR *before* initial average referencing are no no longer permanently flagged as "bad" after initial average referencing, the same as bad-by-deviation/noise/correlation/dropout/RANSAC channels. However, because they're included as part of the `self.unusable_channels` set, they're still always excluded from the calculated average reference signal! This means that initial bad-by-SNR channels are considered good enough to use when interpolating flagged bad channels, but not good enough to include under any circumstances in the average reference. This seems wrong, but is exactly what MATLAB PREP is doing. Any thoughts on whether we want to change this?
    - Currently, bad-by-dropout channels are flagged and removed if detected after initial average referencing (provided `matlab_strict` is False). However, if the signal isn't average-referenced to begin with, wouldn't the initial average referencing make any intermittent flat regions in the data no longer flat? As such, wouldn't we want bad-by-dropout channels to be flagged as unusable *prior* to initial referencing, the same as with bad-by-NaN and bad-by-flat channels?

sappelhoff commented:
> Any thoughts on whether there should be an API for configuring max_iterations at the PrepPipeline level? There's currently no API for configuring NoisyChannels thresholds from PrepPipeline either, so I wasn't sure what the expectation was for Reference settings

I think that if a parameter is configurable in MatPREP, and it's reasonably easy to make it configurable in PyPrep, then we SHOULD make it possible. These params should always have a reasonable default (like in MatPrep) though. On the other hand, let's not expose more params than MatPrep does (for now), unless we judge it very very important.

> This means that initial bad-by-SNR channels are considered good enough to use when interpolating flagged bad channels, but not good enough to include under any circumstances (even if interpolated) in the average reference

🤔 yeah, that sounds off. +1 for an issue about this 1. on pyprep and 2. on matprep. The one for pyprep can be marked to be resolved in version 0.5 --> version 0.4 being targeted more at feature completeness with matprep. If nobody ever reacts to this on matprep, we can do our own, non-matprep-strict thing in pyprep starting with 0.5.

> However, if the signal isn't average-referenced to begin with, wouldn't the initial average referencing make any intermittent flat regions in the data no longer flat? As such, wouldn't we want bad-by-dropout channels to be flagged as unusable prior to initial referencing, the same as with bad-by-NaN and bad-by-flat channels?

Yes, that's true, good point. What does matprep do here?

sappelhoff reviewed (commented):
there are three lines in this diff that are not being covered --> is it possible to reasonably do so with little changes/effort? If not, no problem.

sappelhoff commented on docs/matlab_differences.rst:
```diff
@@ -148,3 +148,43 @@ roughly mean-centered. and will thus produce similar values to normal Pearson
+
+5) Steps 2 through 4 are repeated until either two iterations have passed and
+   no new noisy channels have been detected since the previous iteration, or
+   the maximum number of reference iterations has been exceeded (default: 5).
```
isn't the default 4?

sappelhoff replied:
ah sorry, I looked at an old commit. Ignore this please.

a-hurst commented:
> I think that if a parameter is configurable in MatPREP, and it's reasonably easy to make it configurable in PyPrep, then we SHOULD make it possible.

Okay, I'll add it to the `params` dict for now!

> +1 for an issue about this 1. on pyprep and 2. on matprep. The one for pyprep can be marked to be resolved in version 0.5 --> version 0.4 being targeted more at feature completeness with matprep. If nobody ever reacts to this on matprep, we can do our own, non-matprep-strict thing in pyprep starting with 0.5.

Sounds good, will do! Maybe this is a place for some experimentation, but my gut instinct would be to default to only marking initial bad-by-flat and bad-by-NaN channels as fully "unusable". I'd imagine that the original intent was to removing channels that contribute more noise than signal from contaminating the average reference, but given how sensitive the 'bad-by-correlation' default settings are (if over 1% of all correlation windows are below threshold, the whole channel is considered bad) and how that can change after average referencing, I don't think initial bad-by-SNR channels should be treated as permanently bad. Anyway, that can all be discussed in the new issue.

> Yes, that's true, good point. What does matprep do here?

It just ignores dropout channels completely during referencing: only bad-by-flat/NaN/SNR channels are permanently flagged and removed from the average reference at the onset, and due to the likely bug we've reported already they won't get identified or excluded on subsequent iterations either.

> there are three lines in this diff that are not being covered --> is it possible to reasonably do so with little changes/effort? If not, no problem.

The `matlab_strict` line will get covered in the EEGLAB interpolation PR when I can add the MatPREP comparison tests, but I'll try adding a test case to cover the other main ones!

a-hurst commented:
Okay, I've made codecov a bit happier now: I covered a line that only gets run if the input signal is completely clean, fixed a bug that prevented a ValueError from getting covered (it was throwing a ValueError earlier, so the test passed but the coverage failed), and worked around a known `coverage.py` bug where a `break` immediately after an `if` statement doesn't get counted as being run.

sappelhoff commented:
Thanks @a-hurst
